lucene/core/src/java/org/apache/lucene/search/TopFieldCollector.java

```diff
@@ -150,11 +150,15 @@ public void setScorer(Scorable scorer) throws IOException {
       if (minScoreAcc != null) {
         updateGlobalMinCompetitiveScore(scorer);
       }
-      if (filteringLeafComparator != null && queueFull && hitsThresholdChecker.isThresholdReached()) {
-        // if queue became full and hitsThreshold was reached in previous segments,
-        // notify this segment's leaf comparator that its competitive iterator can be updated
-        filteringLeafComparator.setCanUpdateIterator();
-        totalHitsRelation = TotalHits.Relation.GREATER_THAN_OR_EQUAL_TO;
+
+      if (filteringLeafComparator != null && hitsThresholdChecker.isThresholdReached()) {
+        // hitsThreshold was reached in previous segments, notify this segment's leaf comparator about it
+        filteringLeafComparator.setHitsThresholdReached();
+        // if queue became full in previous segments, notify this segment's leaf comparator about it
+        if (queueFull) filteringLeafComparator.setQueueFull();
+        if (filteringLeafComparator.iteratorUpdated()) {
+          totalHitsRelation = TotalHits.Relation.GREATER_THAN_OR_EQUAL_TO;
+        }
```

Comment:
See my previous comment, I think we should call `setHitsThresholdReached` once, when the threshold is reached. 
Same for `setQueueFull` that is already call in `setBottom`. Said differently, this part could be entirely removed ;).

Reply:
addressed in 746c8fa